### PR TITLE
refactor(fsspec): remove upper bound on fsspec

### PR DESCRIPTION
**Pull request by gforsyth**
We're using `pins` as a dependency in `ibis-framework` and ideally we
can remove the upper-pin on `fsspec` so that doesn't impact our users.
I think the usage of the moved `hash_name` function is the only breaking
change between the current upper pin and most recent release, so I've
added a small workaround to handle the change.

All of the local tests pass, I didn't yet set up sufficient credentials
to handle the various cloud tests.

### CODE REVIEW

**Comment by gforsyth** [comment]
Thanks for taking a look, @isabelizimm !
I haven't been able to replicate the (local) failures, but I'll also try to mess around with my environment and see if I can find out if there's some combination of dependencies that isn't being handled gracefully here.

**Comment by isabelizimm** [comment]
Are you able to check what version of fsspec you are testing on? Locally I'm able to hit these same errors when upgrading to the latest version, `2023.10.0`. It looks like there were some changes to the way `fsspec.LocalFileSystem` is created, which affects how `board_folder` is created in pins. Specifically, there is a change in the name of the `fsspec.fs.protocol` used and there is no longer the attribute `LocalFileSystem._check_file`.

I'm happy to build off your branch and fix those pieces, if you are okay with that!

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -20,14 +20,11 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python: ["3.7", "3.8", "3.9", "3.10", "3.11"]
+        python: ["3.8", "3.9", "3.10", "3.11"]
```
Comment (isabelizimm) [inline .github/workflows/ci.yml:23]: fsspec has stopped supporting python 3.7, which was EOL a few months ago.

**Comment by isabelizimm** [comment]
cc: @machow have you ever seen the `RecursionError: maximum recursion depth exceeded while calling a Python object` error when you were in pins land? All of these tests pass/docs build locally, even with old fsspec versions 😩

File: pins/cache.py [hunk 2]
```diff
@@ -57,12 +57,62 @@ def prefix_cache(fs, board_base_path):
     return f"{proto_name}_{base_hash}"
 
 
+class HashMapper:
```
Comment (isabelizimm) [inline pins/cache.py:60]: rather than having different classes for each cache type, we can use custom mappers

File: pins/cache.py [hunk 3]
```diff
@@ -57,12 +57,62 @@ def prefix_cache(fs, board_base_path):
...
 class PinsCache(SimpleCacheFileSystem):
     protocol = "pinscache"
 
-    def __init__(self, *args, hash_prefix=None, **kwargs):
+    def __init__(self, *args, hash_prefix=None, mapper=HashMapper, **kwargs):
         super().__init__(*args, **kwargs)
         self.hash_prefix = hash_prefix
+        self._mapper = mapper(hash_prefix)
+
+    def hash_name(self, path, *args, **kwargs):
```
Comment (isabelizimm) [inline pins/cache.py:114]: the `hash_name` method is for backwards compatibility with `fsspec<2023.9.0`

File: pins/cache.py [hunk 4]
```diff
@@ -154,15 +175,15 @@ def hash_name(self, path, same_name):
...
+        super().__init__(*args, **kwargs)
+        self.hash_prefix = hash_prefix
+        self._mapper = mapper(hash_prefix)
 
-        # hash full path, and put anything after the final / at the end, just
-        # to make it easier to browse.
-        base_name = super().hash_name(path, same_name)
-        suffix = Path(path).name
-        return f"{base_name}_{suffix}"
+    def hash_name(self, path, *args, **kwargs):
```
Comment (isabelizimm) [inline pins/cache.py:185]: the `hash_name` method is for backwards compatibility with `fsspec<2023.9.0`

File: .github/workflows/ci.yml (hunk 1 above)
Comment (isabelizimm) [reply .github/workflows/ci.yml]: moved this into its own PR for searchability #224

**Comment by isabelizimm** [comment]
Alright, this ended up being some careful cache extraction, but it's in a good place now! Thanks @machow for the pairing sessions to help guide this through. If you have any post-merge questions/comments/concerns, let's make those a follow-up PR 🙌

**Pull request merged**